### PR TITLE
soc: intel_adsp_cavs: store PS when power gating secondary core

### PR DESCRIPTION
**Pull request by kv2019i**
When non-primary core is powered down and restart with sequence of:
 - PM state set to SOFT_OFF
 - once target core is idle, cut power with soc_adsp_halt_cpu()
 - power up core again with k_smp_cpu_resume()

The execution will continue from stored DSP core context, but will hit an assert in z_smp_cpu_mobile() as the PS.INTLEVEL is zero.

Fix this issue by storing and restoring PS register in this flow.

Link: https://github.com/zephyrproject-rtos/zephyr/issues/70181

### CODE REVIEW

**Comment by kv2019i** [comment]
Publishing as draft as I'm not fully confident yet this is the right fix. It seems to fix the test case of https://github.com/zephyrproject-rtos/zephyr/issues/70181 and https://github.com/thesofproject/sof/issues/8942, but what I'm less confident is whether I'm breaking something else. But given we might have multiple people looking at this, pushing an early draft.

**Comment by kv2019i** [comment]
Results look.

Baseline (without this patch):
https://sof-ci.01.org/sofpr/PR8878/build3389/devicetest/index.html

With this patch:
https://sof-ci.01.org/sofpr/PR8944/build3407/devicetest/index.html

Marking as ready for review.

**Review by dcpleung — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Review by andyross — CHANGES_REQUESTED** [review]
This looks wrong (or at least dangerous) to me?  The PS register contains not only the interrupt mask, but also the register window state.   Once you do that WSR.PS, you need to be absolutely sure that all the GPR state all the way up the window stack is exactly the same as it was at the RSR.PS.  That means the functions need to be at exactly the same stack depth and have the same caller state.

It also means that all the spill state needs to be identical: if a frame was in registers (i.e. not the stack) at the time of suspend, it must somehow be restored to those registers on resume.  Otherwise the function will end up rotating back into what amounts to garbage in the register set.  Maybe there's a register spill somewhere else that's taking care of this?  If nothing else it should be documented carefully.

Basically: the commit message says it's doing this to fix the interrupt state.  If that's your goal then mask off the bottom four bits and set them with a RSIL instruction instead (or read+mask+set, etc...), which doesn't touch the dangerous bits of PS.

**Comment by andyross** [comment]
And just to editorialize, and I've made this point before: this ad hoc context save in the platform layer is fragile and dangerous, and clearly is getting more so over time.  Can we please move to a suspend-out-of-the-idle-thread mechanism where the architecture code already provides support?

**Comment by teburd** [comment]
> This looks wrong (or at least dangerous) to me? The PS register contains not only the interrupt mask, but also the register window state. Once you do that WSR.PS, you need to be absolutely sure that all the GPR state all the way up the window stack is exactly the same as it was at the RSR.PS. That means the functions need to be at exactly the same stack depth and have the same caller state.
> 
> It also means that all the spill state needs to be identical: if a frame was in registers (i.e. not the stack) at the time of suspend, it must somehow be restored to those registers on resume. Otherwise the function will end up rotating back into what amounts to garbage in the register set. Maybe there's a register spill somewhere else that's taking care of this? If nothing else it should be documented carefully.
> 
> Basically: the commit message says it's doing this to fix the interrupt state. If that's your goal then mask off the bottom four bits and set them with a RSIL instruction instead (or read+mask+set, etc...), which doesn't touch the dangerous bits of PS.

For secondary cores the resume is meant to effectively start the core from scratch, but this actually doesn't work for a different reason as the idle thread was still running and never reset for that core. Meaning if you go to start the core from scratch again the idle thread appears to be running still and can't be scheduled.

**Comment by andyross** [comment]
> Given that, does it really matter what any of the registers are? 

I'm not following?  Is the code going to execute a "return" and expect that the contents of the registers in the caller (including things like its own return address in A0!) match what was there when the call was made?  Again, absent some deliberate spill code immediately adjacent to the suspend, I don't see how this can ever be made correct.  If it works, it's just by luck; e.g. there is a deep call tree just before the suspend point that already flushed the windows, something like that.

Again, if you want to restore the interrupt mask, the proper mechanism is to set only the bottom four bytes of PS, not the whole register.

**Comment by andyross** [comment]
Sorry, missed this bit:

> Meaning if you go to start the core from scratch again the idle thread appears to be running still and can't be scheduled.

On an SMP system, active threads aren't in the run queue and (obviously) aren't blocked on any wait queues.  There's no state to update when the idle thread suspends[1].  Just reboot, re-initialize the idle thread through the standard path, and let it start running normally on the first context switch.  All the code paths can be the same as the initial boot, which is the intent: SMP on Zephyr is greedy and opportunistic, if you have another CPU you can start it at any time.

[1] Well, there can be bugs.  Like for example it had better not suspend while holding any locks.  But in general there will be no external references to the idle thread, by design.

**Comment by teburd** [comment]
> On an SMP system, active threads aren't in the run queue and (obviously) aren't blocked on any wait queues. There's no state to update when the idle thread suspends[1]. Just reboot, re-initialize the idle thread through the standard path, and let it start running normally on the first context switch. All the code paths can be the same as the initial boot, which is the intent: SMP on Zephyr is greedy and opportunistic, if you have another CPU you can start it at 

Yes, but we don't supply any sensible way of re-initializing the idle thread today was my point.

**Review by tmleman — APPROVED** [review]
LGTM. We have already implemented this for ACE https://github.com/zephyrproject-rtos/zephyr/commit/ff436674975474dd028c38fd335c6813bac51c91.
We can make any further improvements as needed in the future.

**Review by andyross — CHANGES_REQUESTED** [review]
I'm digging in on this one, sorry.  The OWB/CALLINC fields of PS are sensitive to the specific history of the window exception handlers before the save code is reached and simply can't be made safe with the code as it stands (you'd have to do a full spill/reset in the save code).

The commit message indicates that you just need to clear the interrupt mask state, in which case this should work safely (and also be smaller and faster as it requires no support on the suspend path):

`__asm__ volatile("rsil %0, 0" : "=r"(some_dummy_var))`

**Comment by andyross** [comment]
Actually on further thought, this isn't the time for this fight.  If the same technique is already in tree then the damage is done.  Please consider the one-liner fix, as it's a better expression of the bug it's claiming to fix.  If that doesn't work, there may be more problems, in which case I'll show up to I-told-you-so again I guess.  But really this design is going to be a persistent problem on this hardware for as long as it's in there.  Platform code should not be writing its own context switch utilities, and this bug is an example of why.

**Comment by marc-hb** [comment]
@kv2019i just turned off multicore on CAVS:
https://github.com/thesofproject/sof/commit/6bdbf7186ea120827

We needed to get our CI back to green at last, thanks @kv2019i

**Comment by kv2019i** [comment]
@andyross Thanks for reviews. I'm first to admit I didn't fully grok this code. I'm just a/the poor SOF maintainer playing a game of Whac-A-Mole trying keep basic features working on SOF when we update Zephyr. Here I tried to copy the flow from newer Intel platforms.

This is pretty funky code indeed. So summarize my understanding reading the code again today:
 - PM core pm_system_suspend() calls  pm_state_set(SOFT_OFF)
 - the cAVS2.5 code calls "xthal_window_spill()", "_save_core_context()"  and then "k_cpu_atomic_idle(arch_irq_lock());" 
 - core is powered of
 - on resume, we jump to dsp_restore_vector(), call _restore_core_context(), A0/A1 restored 
 - magic bit is the "retw.n" which actually jump back to PM core now from power_gate_exit()
 - when PM core calls pm_state_exit_post_ops(), we actually have a RSIL already to clear PS_INTLEVEL
 
Without my patch, PM core will hit an assert in right after resume due to call
``` 
#define _current_cpu ({ __ASSERT_NO_MSG(!z_smp_cpu_mobile()); \
»       »       »       arch_curr_cpu(); })
```
... and yeah, problem is the interrupts are not locked and thus the assert (sorry, I got this backwards originally). So reset INTLEVEL won't help.

So with above newly acquired understanding, I came up with following alternative fix:
```
--- a/soc/intel/intel_adsp/cavs/power.c
+++ b/soc/intel/intel_adsp/cavs/power.c
@@ -116,7 +131,7 @@ __asm__(".align 4\n\t"
        "dsp_restore_vector:\n\t"
        "  movi  a0, 0\n\t"
        "  movi  a1, 1\n\t"
-       "  movi  a2, 0x40020\n\t"/* PS_UM|PS_WOE */
+       "  movi  a2, 0x4002f\n\t"/* PS_UM|PS_WOE|PS_INTLEVEL(max) */
```

This passes the test, when we return execution to PM core from suspend, PS_INTLEVEL is 0xf and z_smp_cpu_mobile() is false. We reset INTLEVEL in pm_exit_ops, so the above is fine.

Now, this still looks pretty funky to me. Wouldn't it be more safe to save and restore PS so return to PM core via RET.W works as expected (e.g. what @ceolin 's https://github.com/zephyrproject-rtos/zephyr/commit/ff436674975474dd028c38fd335c6813bac51c91 solves). Now we basicly (on Intel cAVS2.5) hardcode PS value that just happens to be sufficient to return from soc/intel/intel_adsp/cavs/power.c:power_gate_exit() to PM core, but the solution on newer Intel platforms seems more solid.

Let me know which is better. Or if we need to wait for bigger changes.

**Comment by andyross** [comment]
Actually this isn't as bad as my initial freakout implied: I was confusing (not for the first time, actually) the PS.OWB field with the actual WINDOWBASE pointer.  The latter is critical, it stores the offset within the register set where the active GPRs are found, and if you modify it on resume you'll end up in the wrong spot, handling the wrong spill/fill exceptions at the wrong registers, and everything will fail.  But WINDOWBASE is a separate SR, not part of PS.

And that's not what OWB is.  OWB is the "Old WINDOWBASE" field.  It's a copy of WINDOWBASE that the hardware makes on entry to a window spill/fill exception.  And since those exceptions aren't themselves interruptible (they run with PS.EXCM=1) the value of those bits at any other time is irrelevant.

So the code as submitted is fine enough; it won't/shouldn't fail.  I still hate the synchronous suspend trickery being played given that the same thing can be done by just booting the core normally and letting it suspend out of idle, but that's a design argument for another time.

**Review by andyross — APPROVED** [review]
(Drop a +1 just to settle the argument.  We can fight about suspend architectures in the abstract future)

**Comment by marc-hb** [comment]
> But really this design is going to be a persistent problem on this hardware for as long as it's in there. Platform code should not be writing its own context switch utilities, and this bug is an example of why.

Does this prevent **testing** this flow? Ideally, using some generic (not platform-specific) test code?

Something like #69365 maybe?

I've been following this from a very safe distance so I have a very vague understanding but the scenarios described above do not look like they require SOF. Do they?

No matter where and how something is implemented:
- if it's not tested then it does not work
- if it's not tested regularly then it will regress soon
- when it's tested, it works only in the configurations tested.

Maybe we've reached a point where getting rid of this platform-specific code would have cost less time than maintaining it.

But for sure we're way past the point where some test code (that can run without SOF) would have saved a massive amount of time.

**Review by nashif — APPROVED** [review]

**Comment by andyross** [comment]
@marc-hb Absolutely +1 to the ability to unit test this.  The difficulty is surely that the host power management support only exists in the SOF driver right now, it would have to be duplicated for a unit test (or have some kind of unit test support added to the kernel).  That's been done for loading, obviously, so the framework is there.  But it's a big task.

**Pull request merged**
